connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamHandler.scala

```diff
@@ -58,10 +59,49 @@ class SparkConnectStreamHandler(responseObserver: StreamObserver[ExecutePlanResp
     // Extract the plan from the request and convert it to a logical plan
     val planner = new SparkConnectPlanner(session)
     val dataframe = Dataset.ofRows(session, planner.transformRelation(request.getPlan.getRoot))
-    processAsArrowBatches(request.getClientId, dataframe)
+    processAsArrowBatches(request.getClientId, dataframe, responseObserver)
+    responseObserver.onNext(sendMetricsToResponse(request.getClientId, dataframe))
+    responseObserver.onCompleted()
+  }
+
+  private def sendMetricsToResponse(clientId: String, rows: DataFrame): ExecutePlanResponse = {
+    // Send a last batch with the metrics
+    ExecutePlanResponse
+      .newBuilder()
+      .setClientId(clientId)
+      .setMetrics(MetricGenerator.buildMetrics(rows.queryExecution.executedPlan))
+      .build()
+  }
+
+  private def handleCommand(session: SparkSession, request: ExecutePlanRequest): Unit = {
+    val command = request.getPlan.getCommand
+    val planner = new SparkConnectPlanner(session)
+    planner.process(command, request.getClientId, responseObserver)
+    responseObserver.onCompleted()
```

Comment:
Not your problem: Commands can have metrics.

Reply:
Added